### PR TITLE
Make PropertyFetcher work for multiple target object types

### PR DESCRIPTION
**Pull request by ejsmith**
The Elasticsearch instrumentation is passing in various different object types as the payload from their event source. When that happens the PropertyFetcher fails because it is caching the innerFetcher based on the first object type it sees. This changes adds a concurrent dictionary of fetchers for any target types that are passed in. I think what the Elasticsearch client is doing is considered bad practice so you may not want to take this change, but I can say though that it is super confusing when this happens and it seems like the property fetcher is randomly failing. Thoughts?

### CODE REVIEW

**Comment by ejsmith** [comment]
Looks like there are some tests failing. I will take a look at those after getting feedback on whether you want this PR or not.

**Comment by CodeBlanch** [comment]
@ejsmith My thoughts... congrats this issue has been in there forever and you are the first person lucky enough to run into it :smile: My concern with this approach is the hash + lookup into the dictionary isn't free. It will introduce a perf hit into all instrumentation using PropertyFetcher perfectly fine up until now.

Do you by chance have a link to the Elasticsearch source where it's creating events? Just curious if there's anything we can key off of. It's not using PropertyFetcher, but there's a similar problem with Activity.TagObjects today. If it is empty, it returns a different type than when it is non-empty. Here's how I handled that: 
https://github.com/open-telemetry/opentelemetry-dotnet/blob/d802472a3661419670e1606ae68e38273dac1b42/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L197-L210

Basically have a special case for empty and then do the reflection magic for non-empty.

**Comment by ejsmith** [comment]
Is there any benchmarks setup that would show what kind of performance impact this change would have? Are you dead set against this change? If so, I will just use my own copy of the PropertyFetcher in my instrumentation. But I am saying though that as things stand, this leads to very confusing behavior that is completely hidden. I know you generally don't want to throw inside of instrumentation type code, but it seems like there should be some sort of way to tell that the property fetch failed. Right now it just silently returns the default value for the type.

**Comment by ejsmith** [comment]
Got the tests passing and code coverage to be happy.

Also, I would guess that the performance of looking up an entry in a very small dictionary is going to be very miniscule compared to the cost of the reflection call to get the property value.

**Comment by CodeBlanch** [comment]
> Also, I would guess that the performance of looking up an entry in a very small dictionary is going to be very miniscule compared to the cost of the reflection call to get the property value.

I think it will be significant. The whole point of the PropertyFetcher is for minuscule performance improvement 😄 

> Are you dead set against this change?

No, I'm just not convinced we can't find a solution that keeps the performance as is. Another thing to consider besides the perf, I think having the type mutating underneath could be equally confusing to people not expecting that.

I was just looking at some of the elasticsearch-net source...

https://github.com/elastic/elasticsearch-net/blob/master/src/Elasticsearch.Net/Diagnostics/DiagnosticSources.cs
https://github.com/elastic/elasticsearch-net/blob/master/src/Elasticsearch.Net/Transport/Pipeline/RequestPipeline.cs

It seems that for a given operationName the payload is always the same. How are you handling this right now, without this change available to contrib repo? Could you have a different PropertyFetcher for each of the types (I think there's 3) and then choose the one to use based on the operation name? Smaller perf hit than the dictionary, and will be local to the Elasticsearch instrumentation.

**Comment by CodeBlanch** [comment]
@ejsmith I put together a quick benchmark in a branch if you want to mess with it: https://github.com/CodeBlanch/opentelemetry-dotnet/blob/propertyfetcher-perf/test/Benchmarks/Instrumentation/PropertyFetcherBenchmark.cs

|                        Method |     Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |---------:|---------:|---------:|-------:|------:|------:|----------:|
|               PropertyFetcher | 11.12 ns | 0.174 ns | 0.163 ns |      - |     - |     - |         - |
| PropertyFetcherWithDictionary | 56.42 ns | 1.154 ns | 1.618 ns | 0.0124 |     - |     - |     104 B |

**Comment by ejsmith** [comment]
Doh! I was just doing the same thing. 😀 I'll take a look.

**Comment by ejsmith** [comment]
Alright, I submitted a new PR in the contrib repo to address this issue. I'll close this.
